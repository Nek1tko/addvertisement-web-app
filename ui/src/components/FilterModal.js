import React, {useEffect, useState} from "react";
import Button from "@mui/material/Button";
import DialogContent from "@mui/material/DialogContent";
import DialogActions from "@mui/material/DialogActions";
import DialogTitle from "@mui/material/DialogTitle";
import Dialog from "@mui/material/Dialog";
import FilterListIcon from "@mui/icons-material/FilterList";
import {Collapse, InputAdornment, MenuItem, TextField} from "@material-ui/core";
import axios from "axios";
import authHeader from "../services/auth-header";
import {Alert} from "@mui/material";

const API_URL = "http://localhost:8080/";
const API_URL_SEARCH = "http://localhost:8080/ad/page";

export const FilterModal = (props) => {
    const [metro, setMetro] = useState('');
    const [metroList, setMetroList] = useState([]);

    const [category, setCategory] = useState('');
    const [categoryList, setCategoryList] = useState([]);

    const [minPrice, setMinPrice] = useState('');
    const [maxPrice, setMaxPrice] = useState('');

    const typeList = [
        {
            id: 0,
            name: "Активно",
        },
        {
            id: 1,
            name: "Неактивно",
        },
        {
            id: 2,
            name: "Все",
        }
    ];
    const [isActive, setActive] = useState('');

    const [alertOpen, setAlertOpen] = useState(false);

    useEffect(() => {
        axios
            .get(API_URL + "metro", {headers: authHeader()})
            .then(res => {
                setMetroList(res.data);
            })
    }, []);

    useEffect(() => {
        axios
            .get(API_URL + "category", {headers: authHeader()})
            .then(res => {
                setCategoryList(res.data);
            })
    }, []);

    useEffect(() => {
        axios
            .get(API_URL + "metro", {headers: authHeader()})
            .then(res => {
                setMetroList(res.data);
            })
    }, []);

    const [open, setOpen] = React.useState(false);
    const handleClickOpen = () => {
        setOpen(true);
    };
    const handleClose = () => {
        setOpen(false);
        setAlertOpen(false);
    };
    const handleApply = () => {
        if (validatePrice()) {
            setOpen(false);
            setAlertOpen(false);
            const jsonFilter = parseRowsToJSON();
            props.setJsonFilter(jsonFilter);
            // postRequest(jsonFilter);
        } else {
            setAlertOpen(true);
        }
    };
    const handleCancel = () => {
        setOpen(false);
        setAlertOpen(false);
    };

    const validatePrice = () => {
        let isCorrect = true;
        if (minPrice !== "" && maxPrice !== "") {
            if (minPrice > maxPrice) {
                isCorrect = false;
            }
        }
        return isCorrect;
    }

    const parseRowsToJSON = () => {
        let jsonModel = {};
        jsonModel["page"] = props.page;
        if (minPrice !== "") {
            jsonModel["minPrice"] = minPrice;
        }
        if (maxPrice !== "") {
            jsonModel["maxPrice"] = maxPrice;
        }
        if (metro !== '') {
            jsonModel["metroId"] = metro;
        }
        if (category !== '') {
            jsonModel["categoryId"] = category;
        }
        if (isActive !== '') {
            if (isActive === "Активно") {
                jsonModel["isActive"] = true;
            } if (isActive === "Неактивно") {
                jsonModel["isActive"] = false;
            }
        }
        console.log(jsonModel);
        props.setJsonFilter(jsonModel);
        return jsonModel;
    }

    const postRequest = (jsonFilter) => {
        axios
            .post(API_URL_SEARCH, {...jsonFilter}, {headers: authHeader()})
            .then(res => {
                props.setAds(res.data);
            })
    }

    return (
        <div>
            <Button
                onClick={handleClickOpen}
                variant="contained"
                startIcon={<FilterListIcon/>}
                style={{
                    backgroundColor: "#FFFFFF",
                    borderColor: "#000000",
                    color: "#000000",
                    marginTop: 20,
                }}
            >
                Фильтры
            </Button>
            <Dialog
                open={open}
                onClose={handleClose}
                onBackdropClick={handleCancel}
                minwidth="xs"
                fullWidth
            >
                <DialogTitle>
                    Фильтры
                </DialogTitle>
                <DialogContent dividers>
                    <Collapse in={alertOpen}>
                        <Alert
                            severity="error"
                            sx={{mb: 2}}
                        >
                            Мин. цена не может быть больше макс. цены!
                        </Alert>
                    </Collapse>
                    <TextField
                        fullWidth
                        select
                        variant="filled"
                        label="Метро"
                        value={metro}
                        onChange={e => {
                            setMetro(e.target.value);
                        }}
                    >
                        {metroList.map(metro => {
                            return (
                                <MenuItem key={metro.id} value={metro.id}>
                                    {metro.name}
                                </MenuItem>);
                        })}
                    </TextField>

                    <TextField
                        fullWidth
                        select
                        variant="filled"
                        label="Категория"
                        value={category}
                        onChange={e => {
                            const value = e.target.value;
                            setCategory(value);
                        }}
                    >
                        {categoryList.map(category => {
                            return (
                                <MenuItem
                                    key={category.id}
                                    value={category.id}
                                >
                                    {category.name}
                                </MenuItem>);
                        })}
                    </TextField>

                    <TextField
                        label="Минимальная цена"
                        variant="filled"
                        type="number"
                        fullWidth
                        InputProps={{
                            endAdornment: <InputAdornment position="end">₽</InputAdornment>,
                        }}
                        onChange={e => {
                            if (e.target.value >= 0) {
                                setMinPrice(e.target.value);
                            }
                        }}
                        value={minPrice}
                    />

                    <TextField
                        label="Максимальная цена"
                        variant="filled"
                        type="number"
                        fullWidth
                        InputProps={{
                            endAdornment: <InputAdornment position="end">₽</InputAdornment>,
                        }}
                        onChange={e => {
                            if (e.target.value >= 0) {
                                setMaxPrice(e.target.value);
                            }
                        }}
                        value={maxPrice}
                    />

                    <TextField
                        fullWidth
                        select
                        variant="filled"
                        label="Тип объявления"
                        value={isActive}
                        onChange={e => {
                            const value = e.target.value;
                            setActive(value);
                        }}
                    >
                        {typeList.map(adType => {
                            return (
                                <MenuItem
                                    key={adType.id}
                                    value={adType.name}
                                >
                                    {adType.name}
                                </MenuItem>);
                        })}
                    </TextField>
                </DialogContent>
                <DialogActions>
                    <Button autoFocus onClick={handleCancel}>
                        Отменить
                    </Button>
                    <Button autoFocus onClick={handleApply}>
                        Применить
                    </Button>
                </DialogActions>
            </Dialog>
        </div>
    );
};
