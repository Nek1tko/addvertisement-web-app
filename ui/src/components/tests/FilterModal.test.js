/**
 * @group unit
 */

import '../FilterModal';
import { FilterModal } from '../FilterModal';
import { mount, shallow } from 'enzyme';
import axios from 'axios';
import React from "react";
import { act } from "react-dom/test-utils";

jest.mock('axios');

const props = {
    page: 1,
    setJsonFilter: jest.fn(),
    ads: [],
};

const eventMaxPrice = {
    target: {
        value: 55
    }
};

const eventMaxPrice2 = {
    target: {
        value: 77
    }
};

const eventMinPrice = {
    target: {
        value: 66
    }
};

const eventCategory = {
    target: {
        value: "Category"
    }
};

const eventAdType = {
    target: {
        value: "Активно"
    }
};

const eventAdType2 = {
    target: {
        value: "Неактивно"
    }
};

const eventMetro = {
    target: {
        value: "Metro"
    }
};

const axiosResponseMetro = { data: [{ id: 'metroId', name: 'Lesnaya' }] };

describe('FilterModalTests', () => {
    let wrapper;
    const setState = jest.fn();

    beforeEach(() => {
        wrapper = shallow(<FilterModal{...props} />);
        axios.get.mockResolvedValue(axiosResponseMetro);
        const useStateSpy = jest.spyOn(React, 'useState');
        useStateSpy.mockImplementation((state) => [state, setState]);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('FilterModalRenderTest', () => {
        mount(<FilterModal{...props} />);
    });

    test('FilterModalSnapshotTest', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('FilterModalClickOpenTest', () => {
        act(() => wrapper.find('#filtersButton').props().onClick());
        expect(setState).toHaveBeenCalledWith(true);
    });

    test('FilterModalClickCloseTest', () => {
        act(() => wrapper.find('#dialogModal').props().onClose());
        expect(setState).toHaveBeenCalledWith(false);
    });

    test('FilterModalClickApplyPositiveTest', () => {
        act(() => wrapper.find('#applyButton').props().onClick());
        expect(setState).toHaveBeenCalledWith(false);
    });

    test('FilterModalClickApplyNegativeTest', () => {
        act(() => wrapper.find('#maxPriceTextField').props().onChange(eventMaxPrice));
        act(() => wrapper.find('#minPriceTextField').props().onChange(eventMinPrice));
        act(() => wrapper.find('#applyButton').props().onClick());
        expect(setState).toHaveBeenCalledTimes(0);
    });

    test('FilterModalClickCancelTest', () => {
        act(() => wrapper.find('#dialogModal').props().onBackdropClick());
        expect(setState).toHaveBeenCalledWith(false);
    });

    test('FilterModalClickClearTest', () => {
        act(() => wrapper.find('#clearButton').props().onClick());
        expect(setState).toHaveBeenCalledTimes(0);
    });

    test('FilterModalChangeCategoryTextFieldTest', () => {
        act(() => wrapper.find('#categoryTextField').props().onChange(eventCategory));
        expect(setState).toHaveBeenCalledTimes(0);
    });

    test('FilterModalChangeAdTypeTextFieldTest', () => {
        act(() => wrapper.find('#adTypeTextField').props().onChange(eventAdType));
        expect(setState).toHaveBeenCalledTimes(0);
    });

    test('FilterModalChangeMetroTextFieldTest', () => {
        act(() => wrapper.find('#metroTextField').props().onChange(eventMetro));
        expect(setState).toHaveBeenCalledTimes(0);
    });

    test('FilterModalJsonParseActiveTest', () => {
        act(() => wrapper.find('#maxPriceTextField').props().onChange(eventMaxPrice2));
        act(() => wrapper.find('#minPriceTextField').props().onChange(eventMinPrice));
        act(() => wrapper.find('#categoryTextField').props().onChange(eventCategory));
        act(() => wrapper.find('#metroTextField').props().onChange(eventMetro));
        act(() => wrapper.find('#adTypeTextField').props().onChange(eventAdType));

        act(() => wrapper.find('#applyButton').props().onClick());

        expect(setState).toHaveBeenCalledWith(false);
        expect(props.setJsonFilter).toHaveBeenCalledTimes(2);
    });

    test('FilterModalJsonParseNegativeTest', () => {
        act(() => wrapper.find('#maxPriceTextField').props().onChange(eventMaxPrice2));
        act(() => wrapper.find('#minPriceTextField').props().onChange(eventMinPrice));
        act(() => wrapper.find('#categoryTextField').props().onChange(eventCategory));
        act(() => wrapper.find('#metroTextField').props().onChange(eventMetro));
        act(() => wrapper.find('#adTypeTextField').props().onChange(eventAdType2));

        act(() => wrapper.find('#applyButton').props().onClick());

        expect(setState).toHaveBeenCalledWith(false);
        expect(props.setJsonFilter).toHaveBeenCalledTimes(2);
    });
});
