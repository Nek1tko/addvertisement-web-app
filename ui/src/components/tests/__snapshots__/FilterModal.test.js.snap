// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterModalTests FilterModalSnapshotTest 1`] = `ShallowWrapper {}`;
